Source/ObjectiveDropboxOfficial/Shared/Handwritten/Networking/DBURLSessionTaskWithTokenRefresh.m

```diff
@@ -0,0 +1,235 @@
+  });
+}
+
+- (void)resume {
+  dispatch_async(_serialQueue, ^{
+    self->_started = YES;
+    if (self->_sessionTask) {
+      [self->_sessionTask resume];
+    } else {
+      [self db_start];
+    }
+  });
+}
+
+- (void)setProgressBlock:(DBProgressBlock)progressBlock
+                   queue:(NSOperationQueue *)queue {
+  dispatch_async(_serialQueue, ^{
+    self->_progressBlock = progressBlock;
+    self->_progressQueue = queue;
+    [self db_setProgressHandlerIfNecessary];
+  });
+}
+
+- (void)setResponseBlock:(DBURLSessionTaskResponseBlock *)responseBlock
+                   queue:(NSOperationQueue *)queue {
+  dispatch_async(_serialQueue, ^{
+    self->_responseBlock = responseBlock;
+    self->_responseQueue = queue;
+    [self db_setResponseHandlerIfNecessary];
+  });
+}
+
+#pragma mark Private helpers
+
+- (void)db_start {
+  DBOAuthCompletion completion = ^(DBOAuthResult *result) {
+    dispatch_async(self->_serialQueue, ^{
+      [self db_handleTokenRefreshResult:result];
+    });
+  };
+
+  if (_tokenProvider) {
+    [_tokenProvider refreshAccessTokenIfNecessary:completion];
+  } else {
+    completion(nil);
+  }
+}
+
+- (void)db_handleTokenRefreshResult:(DBOAuthResult *)result {
+  if ([result isError] && result.errorType != DBAuthInvalidGrant) {
+    // Refresh failed, due to an error that's not invalid grant, e.g. A refresh request timed out.
+    // Complete request with error immediately, so developers could retry and get access token refreshed.
+    // Otherwise, the API request may proceed with an expired access token which would lead to
+    // a false positive auth error.
+    [self db_completeWithError:result.nsError];
+  } else {
+    // Refresh succeeded or a refresh is not required, i.e. access token is valid, continue request normally.
+    // Or
+    // Refresh failed due to invalid grant, e.g. refresh token revoked by user.
+    // Continue, and the API call would failed with an auth error that developers can handle properly.
+    // e.g. Sign out the user upon auth error.
+    [self db_initializeSessionTask];
+  }
+}
+
+- (void)db_initializeSessionTask {
+  _sessionTask = _taskCreationBlock();
+  [self db_setProgressHandlerIfNecessary];
+  [self db_setResponseHandlerIfNecessary];
+  if (_cancelled) {
+    [_sessionTask cancel];
+  } else if (_suspended) {
+    [_sessionTask suspend];
+  } else if (_started) {
+    [_sessionTask resume];
+  }
+}
+
+- (void)db_setProgressHandlerIfNecessary {
+  if (_sessionTask && _progressBlock) {
+    [_taskDelegate addProgressHandlerForTaskWithIdentifier:_sessionTask.taskIdentifier
+                                                   session:_session
+                                           progressHandler:_progressBlock
+                                      progressHandlerQueue:_progressQueue];
+  }
+}
+
+- (void)db_setResponseHandlerIfNecessary {
+  if (_sessionTask == nil || _responseBlock == nil) {
+    return;
+  }
+
+  if ([_responseBlock isRpcResponseBlock]) {
```

Comment:
I think it'd be clearer to do `_responseBlock.rpcResponseBlock != nil` given the context. (e.g. your only adding the response block handler to the delegate if it exists. )